enhancements/cguConditionEnhancement.md

```diff
@@ -0,0 +1,452 @@
+- [Kubernetes recommendation on Conditions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties)
+- Posted a question in slack channel : #forum-apiserver for any existing design guidelines. Still waiting for some feedback.
+- **Openshift-apiserver, kube-apiserver, machine-config, node-tuning, config-operator** have below conditions:
+
+```yaml
+[root@hv2 ~]# oc get co/openshift-apiserver -oyaml | yq '.status.conditions'
+[
+  {
+    "lastTransitionTime": "2022-03-21T14:38:09Z",
+    "message": "All is well",
+    "reason": "AsExpected",
+    "status": "False",
+    "type": "Degraded"
+  },
+  {
+    "lastTransitionTime": "2022-03-21T14:39:58Z",
+    "message": "All is well",
+    "reason": "AsExpected",
+    "status": "False",
+    "type": "Progressing"
+  },
+  {
+    "lastTransitionTime": "2022-03-21T14:43:16Z",
+    "message": "All is well",
+    "reason": "AsExpected",
+    "status": "True",
+    "type": "Available"
+  },
+  {
+    "lastTransitionTime": "2022-03-21T14:38:08Z",
+    "message": "All is well",
+    "reason": "AsExpected",
+    "status": "True",
+    "type": "Upgradeable"
+  }
+]
+```
+
+- **Network operator** has below condition:
+
+```yaml
+[root@hv2 ~]# oc get co/network -ojson | jq '.status.conditions'
+[
+  {
+    "lastTransitionTime": "2022-03-21T14:35:06Z",
+    "status": "False",
+    "type": "Degraded"
+  },
+  {
+    "lastTransitionTime": "2022-03-21T14:35:06Z",
+    "status": "False",
+    "type": "ManagementStateDegraded"
+  },
+  {
+    "lastTransitionTime": "2022-03-21T14:35:06Z",
+    "status": "True",
+    "type": "Upgradeable"
+  },
+  {
+    "lastTransitionTime": "2022-06-07T15:35:25Z",
+    "status": "False",
+    "type": "Progressing"
+  },
+  {
+    "lastTransitionTime": "2022-03-21T14:37:57Z",
+    "status": "True",
+    "type": "Available"
+  }
+]
+```
+
+## Proposal
+
+This section proposes minimalistic changes without breaking the current design pattern. The below proposal/recommendation consists of more granular `status.conditions`  based on all the verification we already have in TALM.
+
+**Considering**: backup,precaching is true and enable for upgrade is true as well. 
+
+### Workflow Description
+
+1. Condition type related to cluster existing in the managedcluster. The type could reflect on the existence of the targeted cluster to be present in the managedcluster. 
+
+  ```yaml
+  Type:    "ClusterValidation/ClustersSelected",
+  Status:  True/False,
+  Reason:  "",
+  Message: “”
+  ```
+
+2. Condition type: **BackupDone**
+3. a) Spec validation for the `managed policies`. Currently it gets checked both during pre-caching and for upgrade. This check can be done before starting an upgrade, irrespective of pre-caching being true or false. b) Platform related verification. 
```

Comment:
I kept preupgrade validation after backup, because the idea is to first do the backup done. In addition, backup task is not dependent on any kind of preupgrade/upgrade/precaching validation. Therefore, if in the CGU, backup is set to true, we should take the backup irrespective to any missing managed policy, or invalid platform image is provided. wdyt?

Reply:
Isn't backup supposed to be done right before upgrade starts (or as close as we can)?

Reply:
Is there a case where the backup could be done then the validation condition failed stuck for good time then the upgrade happen ? In that case do we need to re-do the backup ?

Reply:
@jc-rh, @serngawy: in case, the enable field is false, the upgrade never starts if i am not wrong, in that case if a admin creates a cgu with backup: true and enable: false, it creates backup but waits for enable to be true. I don't mind moving preUpgradevalidation before backup.

@serngawy: backup is done only once, since the backup status will always be returned succeeded if it was succeeded earlier. Backup is never triggered more than once

Reply:
it make sense to trigger the backup only once but the timing matter. It should be just before starting the upgrade. simple scenario as mentioned above; backup has done then stuck in validation while the maintenance window time pass then wait for next maintenance time cycle to start upgrade that will make backup older than the time upgrade will happen. 
AND if there is no upgrade just backup it will still work

Reply:
okay

Reply:
in ideal scenario, 12-15 seconds per cluster for the job, since it is with managedclusterAction and monitored by managedclusterView in every 20 seconds interval and reconcile happens in every 30 seconds, so backup can be done in min 60s - max 240s from my current experience.